### PR TITLE
Allow fallback to use formatter for default format

### PR DESCRIPTION
**Pull request by chrisbloom7**
Fix the fallback formatter issue reported in issue #33

### CODE REVIEW

**Comment by dblock** [comment]
Bump @chrisbloom7.

**Comment by akicho8** [comment]
:+1:
I was also troubled with the same problem.

**Comment by chrisbloom7** [comment]
@dblock I'll make a note to add tests. Do I just add an entry at the top of the Changelog?

**Comment by dblock** [comment]
Yes, please @chrisbloom7.

**Comment by dblock** [comment]
Bump @chrisbloom7 ?

**Comment by dblock** [comment]
It's still a real problem. You sure you don't want to finish this?

**Comment by chrisbloom7** [comment]
Oh, really? I figured at 5 years old it would have been addressed somewhere along the line. I haven't used Grape in a while, but if you think the fix for this is still relatively the same in the latest version then I can try to find some time to push it through.

**Comment by dblock** [comment]
Yes, I think it's still a problem and a worthy fix/addition.

**Comment by chrisbloom7** [comment]
@dblock tests added and changelog updated

File: lib/grape-rabl/formatter.rb [hunk 1]
```diff
@@ -29,12 +29,17 @@ def render
...
-          Grape::Formatter::Json.call object, env
+          fallback_formatter.call object, env
         end
       end
 
       private
 
+      def fallback_formatter
+        return @fallback_formatter if defined?(@fallback_formatter)
+        @fallback_formatter = Grape::Formatter.formatter_for(env['api.format']) || Grape::Formatter::Json
```
Comment (dblock) [inline lib/grape-rabl/formatter.rb]: 1) Is there a reason to do the whole return and check? `@fallack_formatter ||= ...` will do just as well, simpler. 
2) It's only used once, do we need to retain a reference to it? I think this method can just return without memoizing the formatter.

File: spec/grape_rabl_formatter.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      expect(last_response.body).to eq(%(<?xml version="1.0" encoding="UTF-8"?>
+<hash>
+<errors type="array">
+<error>bad</error>
+<error>things</error>
+<error>happened</error>
+</errors>
+</hash>
+))
+    end
```
Comment (dblock) [inline spec/grape_rabl_formatter.rb]: The test works for the left side of `Grape::Formatter.formatter_for(env['api.format']) || ...`. You also need a test for the JSON default (right side).

**Review by dblock — CHANGES_REQUESTED** [review]
Thanks for picking this up! 

Please also update the README explaining how the formatter falls back.

Since it's not a backwards compatible change, add an UPGRADING file like https://github.com/ruby-grape/grape/blob/master/UPGRADING.md with a reference to it in README and document how the behavior changed.

Bump the version to 0.5. 

Minor test request below.

File: lib/grape-rabl/formatter.rb (hunk 1 above)
Comment (chrisbloom7) [reply lib/grape-rabl/formatter.rb]: If it's only used once then definitely not. I'll get rid of the memoization. (It was initially three lines which made the early return more useful)

**Comment by chrisbloom7** [comment]
@dblock I'm digging a little deeper into what `env['api.format']` represents and it turns out it's not always going to be the value of `default_format`. According to https://github.com/ruby-grape/grape#api-formats:

> The response format (and thus the automatic serialization) is determined in the following order:
> - Use the file extension, if specified. If the file is `.json`, choose the JSON format.
> - Use the value of the `format` parameter in the query string, if specified.
> - Use the format set by the `format` option, if specified.
> - Attempt to find an acceptable format from the `Accept` header.
> - Use the default format, if specified by the `default_format` option.
> - Default to `:txt`

I think it still makes sense to use `env['api.format']` to determine what to fall back to. I'll expand the tests to ensure these are all exercised and update the CHANGELOG and README accordingly to say that it's based on the default response as determined by Grape as opposed to the `default_format` option specifically.

File: spec/grape_rabl_formatter.rb (hunk 2 above)
Comment (chrisbloom7) [reply spec/grape_rabl_formatter.rb]: Turns out we'd never hit the JSON default so I've removed it

**Comment by dblock** [comment]
Merged, this is perfect.

@chrisbloom7 Want to join the (co)maintainers group and make a release? Shoot me an email to dblock at dblock dot org with your rubygems email if you're into it.

**Pull request merged**